opal/mca/smsc/xpmem/smsc_xpmem_module.c

```diff
@@ -46,131 +48,160 @@ mca_smsc_endpoint_t *mca_smsc_xpmem_get_endpoint(opal_proc_t *peer_proc)
-        xpmem_ctx->reg[0] = NULL;
-        return OPAL_SUCCESS;
+    if(OPAL_UNLIKELY(-1 == endpoint->apid)) {
+        OBJ_RELEASE(endpoint->vma_module);
+        OBJ_RELEASE(endpoint);
+        return NULL;
     }
 
-    /* let the caller know we found an overlapping registration that can be coalesced into
-     * the requested interval. the caller will remove the last reference and delete the
-     * registration. */
-    return 2;
+    return &endpoint->super;
 }
 
-/* look up the remote pointer in the peer rcache and attach if
- * necessary */
+/* look up the remote pointer in the peer rcache and attach if necessary */
 void *mca_smsc_xpmem_map_peer_region(mca_smsc_endpoint_t *endpoint, uint64_t flags,
                                      void *remote_ptr, size_t size, void **local_ptr)
 {
     mca_smsc_xpmem_endpoint_t *xpmem_endpoint = (mca_smsc_xpmem_endpoint_t *) endpoint;
-    mca_rcache_base_vma_module_t *vma_module = mca_smsc_xpmem_module.vma_module;
-    uint64_t attach_align = 1 << mca_smsc_xpmem_component.log_attach_align;
+    mca_rcache_base_vma_module_t *vma_module = xpmem_endpoint->vma_module;
+    uintptr_t attach_align = 1 << mca_smsc_xpmem_component.log_attach_align;
     mca_rcache_base_registration_t *reg = NULL;
-    mca_smsc_xpmem_check_reg_ctx_t check_ctx = {.endpoint = xpmem_endpoint, .reg = &reg};
     xpmem_addr_t xpmem_addr;
     uintptr_t base, bound;
     int rc;
 
+    // base is the first byte of the region, bound is the last (inclusive)
     base = OPAL_DOWN_ALIGN((uintptr_t) remote_ptr, attach_align, uintptr_t);
-    bound = OPAL_ALIGN((uintptr_t) remote_ptr + size, attach_align, uintptr_t);
+    bound = OPAL_ALIGN((uintptr_t) remote_ptr + size, attach_align, uintptr_t) - 1;
     if (OPAL_UNLIKELY(bound > xpmem_endpoint->address_max)) {
         bound = xpmem_endpoint->address_max;
     }
 
-    check_ctx.base = base;
-    check_ctx.bound = bound;
-
-    /* several segments may match the base pointer */
-    rc = mca_rcache_base_vma_iterate(vma_module, (void *) base, bound - base, true,
-                                     mca_smsc_xpmem_check_reg, &check_ctx);
-    if (2 == rc) {
-        bound = bound < (uintptr_t) reg->bound ? (uintptr_t) reg->bound : bound;
-        base = base > (uintptr_t) reg->base ? (uintptr_t) reg->base : base;
-        mca_smsc_xpmem_unmap_peer_region(reg);
-        reg = NULL;
+    rc = mca_rcache_base_vma_find(vma_module, (void *) base, bound - base + 1, &reg);
+    assert(OPAL_SUCCESS == rc);
+
+    if (reg) {
+        int32_t old_ref_count = opal_atomic_fetch_add_32(&reg->ref_count, 1);
+
+        if (0 == old_ref_count) {
+            /* Registration is being deleted by another thread
+             * in mca_smsc_xpmem_unmap_peer_region, ignore it. */
+            reg = NULL;
+        }
+    } else {
+        /* If there is a registration that overlaps with the requested range, but
```

Comment:
It is possible there are multiple regions that overlaps with the current one (imagine two prior registrations with a gap in the middle, gap that is then covered by this registration). If I understand the code correctly, you expect to merge these 3 registration, but the current code can only merge two registrations (so it will merge one of the earlier registrations with the current one).

Reply:
Indeed I see. Apparently also the old code only merges with the first overlapping area found. I updated the PR, and now the loop does not stop on the first overlapping area it finds, but keeps going and includes potentially all 4 surrounding areas. Please review again.

Reply:
Better, but still suboptimal. There are two issues I noticed:

1.  each `find` operation takes and release a lock on the underlying `opal_interval_tree_t` object. If the merge operation is lowered into the interval object itself, we can do the search and the merge with only one lock and one unlock.
2. having the search outside the `opal_interval_tree_t` also trigger the tree traversal for each `find` operation. If we move the merge down into this object, we can do all the merges with a single tree traversal (because the nodes are ordered).

Reply:
I agree it is suboptimal; we need multiple operations as find() does not support partial matching. There are also some thread safety concerns (in this spot and also in some other ones). It is my intution that to address these concerns the atomicity point needs to be moved inside the tree instead of all the way out in smsc.

Moving the merge operation inside the tree indeed sounds good (and would/could also improve the atomicity thing). Even better if we would have an XPMEM reg immediately ready to take the place of the old ones as soon as we invalidate them.

Anyway besides the rambling above, these problems exist in the old implementation with iterate() and they remain in the find() one suggested here. I don't currently have a concrete plan to tackle them, so I would suggest we merge the find() improvement which is ready and allegedly does not introduce *new* shortcomings. We could create an issue to track fixes and improvements that need to be made in the future.